From 2d8c2694de2e21f47035e1a0651cb3ec641bb60b Mon Sep 17 00:00:00 2001
From: Michael Zimmermann <sigmaepsilon92@gmail.com>
Date: Wed, 26 May 2021 22:30:40 +0200
Subject: [PATCH 1/2] add support for workerBinary configuration

With self-hosted runners all jobs usually have full access to the runner
directory. You can put the runner in a VM or container and use `--once` to
restart with a clean environment after each job but that doesn't prevent jobs
from doing either of the following:
- access runner secrets like `.credentials_rsaparams` to then take over the
  runner
- modify the listener to nuke `--once` and manipulate future jobs.

To fix that issue I'm running `Runner.Listener` in the host environment while
running `Runner.Worker` in a VM or container.
That is easily possible because the only way of communication between the two
are unnamed pipes which can easily be forwarded into another environment.

In my specific setup I set the config `workerBinary` to an absolute path outside
of the runner installation. That binary then starts a qemu VM, forwards the
pipes into it, places a copy of the runner installation into the VM and starts
`Runner.Worker` with the forwarded pipes as the arguments.
---
 src/Runner.Common/ConfigurationStore.cs                   | 3 +++
 src/Runner.Common/Constants.cs                            | 1 +
 src/Runner.Listener/CommandSettings.cs                    | 8 +++++++-
 src/Runner.Listener/Configuration/ConfigurationManager.cs | 2 ++
 src/Runner.Listener/JobDispatcher.cs                      | 5 +++++
 5 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/src/Runner.Common/ConfigurationStore.cs b/src/Runner.Common/ConfigurationStore.cs
index 8528e509..ab0cafeb 100644
--- a/src/Runner.Common/ConfigurationStore.cs
+++ b/src/Runner.Common/ConfigurationStore.cs
@@ -56,6 +56,9 @@ namespace GitHub.Runner.Common
         [DataMember(EmitDefaultValue = false)]
         public string ServerUrlV2 { get; set; }
 
+        [DataMember(EmitDefaultValue = false)]
+        public string WorkerBinary { get; set; }
+
         [IgnoreDataMember]
         public bool IsHostedServer
         {
diff --git a/src/Runner.Common/Constants.cs b/src/Runner.Common/Constants.cs
index e4d2cfc0..cf5806c3 100644
--- a/src/Runner.Common/Constants.cs
+++ b/src/Runner.Common/Constants.cs
@@ -103,6 +103,7 @@ namespace GitHub.Runner.Common
                     public static readonly string UserName = "username";
                     public static readonly string WindowsLogonAccount = "windowslogonaccount";
                     public static readonly string Work = "work";
+                    public static readonly string WorkerBinary = "workerbinary";
 
                     // Secret args. Must be added to the "Secrets" getter as well.
                     public static readonly string Token = "token";
diff --git a/src/Runner.Listener/CommandSettings.cs b/src/Runner.Listener/CommandSettings.cs
index 7c29e2b6..d81f8e73 100644
--- a/src/Runner.Listener/CommandSettings.cs
+++ b/src/Runner.Listener/CommandSettings.cs
@@ -50,7 +50,8 @@ namespace GitHub.Runner.Listener
                     Constants.Runner.CommandLine.Args.UserName,
                     Constants.Runner.CommandLine.Args.WindowsLogonAccount,
                     Constants.Runner.CommandLine.Args.WindowsLogonPassword,
-                    Constants.Runner.CommandLine.Args.Work
+                    Constants.Runner.CommandLine.Args.Work,
+                    Constants.Runner.CommandLine.Args.WorkerBinary
                 },
             // Valid remove flags and args
             [Constants.Runner.CommandLine.Commands.Remove] =
@@ -335,6 +336,11 @@ namespace GitHub.Runner.Listener
             return GetArg(Constants.Runner.CommandLine.Args.MonitorSocketAddress);
         }
 
+        public string GetWorkerBinary()
+        {
+            return GetArg(Constants.Runner.CommandLine.Args.WorkerBinary);
+        }
+
         // This is used to find out the source from where the Runner.Listener.exe was launched at the time of run
         public string GetStartupType()
         {
diff --git a/src/Runner.Listener/Configuration/ConfigurationManager.cs b/src/Runner.Listener/Configuration/ConfigurationManager.cs
index 05b93cb5..97addd96 100644
--- a/src/Runner.Listener/Configuration/ConfigurationManager.cs
+++ b/src/Runner.Listener/Configuration/ConfigurationManager.cs
@@ -411,6 +411,8 @@ namespace GitHub.Runner.Listener.Configuration
 
             runnerSettings.MonitorSocketAddress = command.GetMonitorSocketAddress();
 
+            runnerSettings.WorkerBinary = command.GetWorkerBinary();
+
             _store.SaveSettings(runnerSettings);
 
             _term.WriteLine();
diff --git a/src/Runner.Listener/JobDispatcher.cs b/src/Runner.Listener/JobDispatcher.cs
index f1b4af9e..c31835ca 100644
--- a/src/Runner.Listener/JobDispatcher.cs
+++ b/src/Runner.Listener/JobDispatcher.cs
@@ -45,6 +45,7 @@ namespace GitHub.Runner.Listener
 
         IConfigurationStore _configurationStore;
 
+        private string _workerBinary;
         RunnerSettings _runnerSettings;
         private static readonly string _workerProcessName = $"Runner.Worker{IOUtil.ExeExtension}";
 
@@ -70,6 +71,7 @@ namespace GitHub.Runner.Listener
             _configurationStore = hostContext.GetService<IConfigurationStore>();
             _runnerSettings = _configurationStore.GetSettings();
             _poolId = _runnerSettings.PoolId;
+            _workerBinary = _runnerSettings.WorkerBinary;
 
             int channelTimeoutSeconds;
             if (!int.TryParse(Environment.GetEnvironmentVariable("GITHUB_ACTIONS_RUNNER_CHANNEL_TIMEOUT") ?? string.Empty, out channelTimeoutSeconds))
@@ -466,6 +468,9 @@ namespace GitHub.Runner.Listener
                                 HostContext.WritePerfCounter("StartingWorkerProcess");
                                 var assemblyDirectory = HostContext.GetDirectory(WellKnownDirectory.Bin);
                                 string workerFileName = Path.Combine(assemblyDirectory, _workerProcessName);
+                                if (!String.IsNullOrEmpty(_workerBinary)) {
+                                    workerFileName = _workerBinary;
+                                }
                                 workerProcessTask = processInvoker.ExecuteAsync(
                                     workingDirectory: assemblyDirectory,
                                     fileName: workerFileName,
-- 
2.41.0

